File: scripts/sync-doctors.mjs
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";

const root = path.resolve(new URL("..", import.meta.url).pathname);
const outputPath = path.resolve(process.env.DOCTORS_SYNC_OUTPUT || path.join(root, "data", "doctors.json"));
const localSource = process.env.DOCTORS_SYNC_SOURCE_FILE;
const allowDoctorDrop = process.env.ALLOW_DOCTOR_DROP === "true";
const sheetRange = process.env.GOOGLE_SHEET_RANGE || "Médecins!A:Z";

const columns = {
  id: ["id"],
  lastName: ["nom", "last name", "lastname"],
  firstName: ["prénom", "prenom", "first name", "firstname"],
  title: ["titre", "title"],
  specialty: ["spécialité", "specialite", "specialty"],
  subSpecialty: ["sous-spécialité", "sous-specialite", "sub specialty", "subspecialty"],
  city: ["ville", "city"],
  district: ["quartier", "district"],
  address: ["adresse", "address"],
  phone: ["téléphone", "telephone", "phone"],
  whatsapp: ["whatsapp", "wa"],
  googleMaps: ["google maps", "google map", "maps", "itinéraire", "itineraire"],
  googleRating: ["note google", "google rating", "rating"],
  googleReviews: ["nombre d’avis", "nombre d'avis", "avis google", "google reviews", "reviews"],
  hours: ["horaires", "hours"],
  languages: ["langues", "languages"],
  image: ["photo", "image"],
  status: ["statut", "status"],
  verified: ["vérifié", "verifie", "verified"],
  featured: ["mis en avant", "featured"],
  updatedAt: ["dernière vérification", "derniere verification", "updated_at", "updated at"]
};

function normalizeHeader(value) {
  return String(value || "")
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .replace(/[’']/g, "'")
    .replace(/\s+/g, " ");
}

function buildHeaderMap(headerRow) {
  const normalized = headerRow.map(normalizeHeader);
  const map = {};
  for (const [field, aliases] of Object.entries(columns)) {
    const aliasSet = aliases.map(normalizeHeader);
    map[field] = normalized.findIndex((header) => aliasSet.includes(header));
  }
  return map;
}

function get(row, headerMap, field) {
  const index = headerMap[field];
  return index >= 0 ? clean(row[index]) : "";
}

function clean(value) {
  return String(value ?? "").replace(/\s+/g, " ").trim();
}

function slugify(value) {
  return clean(value)
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
}

function normalizePhone(value) {
  return clean(value)
    .replace(/[().]/g, " ")
    .replace(/\s*[/|;]\s*/g, " / ")
    .replace(/\s+/g, " ");
}

function phoneList(value) {
  return normalizePhone(value)
    .split(/\s*\/\s*|\s*;\s*|\s*\|\s*/)
    .map(clean)
    .filter(Boolean);
}

function toBoolean(value) {
  const normalized = normalizeHeader(value);
  if (["oui", "yes", "true", "vrai", "1", "x"].includes(normalized)) return true;
  if (["non", "no", "false", "faux", "0"].includes(normalized)) return false;
  return false;
}

function toNumber(value) {
  const cleaned = clean(value).replace(",", ".");
  if (!cleaned) return null;
  const number = Number(cleaned);
  return Number.isFinite(number) ? number : null;
}

function toInteger(value) {
  const number = toNumber(value);
  return number === null ? null : Math.trunc(number);
}

function toList(value) {
  return clean(value)
    .split(/\s*[,;|]\s*/)
    .map(clean)
    .filter(Boolean);
}

function normalizeStatus(value) {
  const status = normalizeHeader(value);
  if (["active", "actif", "publie", "publiee", "published"].includes(status)) return "active";
  if (["inactive", "inactif", "masque", "masquee", "hidden"].includes(status)) return "inactive";
  if (["draft", "brouillon", "a verifier", "a-verifier"].includes(status)) return "draft";
  return status ? "draft" : "draft";
}

function normalizeRows(rows) {
  if (!Array.isArray(rows) || rows.length === 0) {
    throw new Error("Le Google Sheet ne contient aucune ligne.");
  }

  const headerMap = buildHeaderMap(rows[0]);
  const doctors = rows.slice(1)
    .filter((row) => row.some((cell) => clean(cell)))
    .map((row, index) => {
      const firstName = get(row, headerMap, "firstName");
      const lastName = get(row, headerMap, "lastName");
      const title = get(row, headerMap, "title") || "Dr";
      const fullName = clean([title, firstName, lastName].filter(Boolean).join(" "));
      const explicitId = get(row, headerMap, "id");
      const phone = normalizePhone(get(row, headerMap, "phone"));
      const id = explicitId || slugify(fullName || `doctor-${index + 1}`);

      if (!id) throw new Error(`Ligne ${index + 2}: ID ou nom obligatoire.`);
      if (!fullName) throw new Error(`Ligne ${index + 2}: nom/prénom/titre manquant.`);

      return {
        id,
        name: fullName,
        title,
        specialty: get(row, headerMap, "specialty"),
        sub_specialty: get(row, headerMap, "subSpecialty"),
        city: get(row, headerMap, "city") || "Kénitra",
        district: get(row, headerMap, "district"),
        address: get(row, headerMap, "address"),
        phone,
        phones: phoneList(phone),
        whatsapp: normalizePhone(get(row, headerMap, "whatsapp")),
        google_maps: get(row, headerMap, "googleMaps"),
        google_rating: toNumber(get(row, headerMap, "googleRating")),
        google_reviews: toInteger(get(row, headerMap, "googleReviews")),
        hours: get(row, headerMap, "hours"),
        languages: toList(get(row, headerMap, "languages")),
        image: get(row, headerMap, "image"),
        status: normalizeStatus(get(row, headerMap, "status")),
        verified: toBoolean(get(row, headerMap, "verified")),
        featured: toBoolean(get(row, headerMap, "featured")),
        updated_at: get(row, headerMap, "updatedAt")
      };
    });

  const ids = new Set();
  for (const doctor of doctors) {
    if (ids.has(doctor.id)) throw new Error(`ID médecin dupliqué: ${doctor.id}`);
    ids.add(doctor.id);
  }

  return doctors;
}

async function getGoogleToken(serviceAccount) {
  const now = Math.floor(Date.now() / 1000);
  const header = Buffer.from(JSON.stringify({ alg: "RS256", typ: "JWT" })).toString("base64url");
  const payload = Buffer.from(JSON.stringify({
    iss: serviceAccount.client_email,
    scope: "https://www.googleapis.com/auth/spreadsheets.readonly",
    aud: "https://oauth2.googleapis.com/token",
    exp: now + 3600,
    iat: now
  })).toString("base64url");
  const signature = crypto
    .createSign("RSA-SHA256")
    .update(`${header}.${payload}`)
    .sign(serviceAccount.private_key, "base64url");

  const response = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
      assertion: `${header}.${payload}.${signature}`
    })
  });

  if (!response.ok) {
    throw new Error(`Google OAuth a échoué: ${response.status} ${await response.text()}`);
  }

  const token = await response.json();
  return token.access_token;
}

async function readGoogleSheet() {
  const sheetId = process.env.GOOGLE_SHEETS_ID;
  const serviceAccountJson = process.env.GOOGLE_SERVICE_ACCOUNT_JSON;
  if (!sheetId) throw new Error("GOOGLE_SHEETS_ID est manquant.");
  if (!serviceAccountJson) throw new Error("GOOGLE_SERVICE_ACCOUNT_JSON est manquant.");

  const serviceAccount = JSON.parse(serviceAccountJson);
  const token = await getGoogleToken(serviceAccount);
  const url = new URL(`https://sheets.googleapis.com/v4/spreadsheets/${sheetId}/values/${encodeURIComponent(sheetRange)}`);
  url.searchParams.set("majorDimension", "ROWS");

  const response = await fetch(url, { headers: { authorization: `Bearer ${token}` } });
  if (!response.ok) {
    throw new Error(`Lecture Google Sheets impossible: ${response.status} ${await response.text()}`);
  }

  const data = await response.json();
  return data.values || [];
}

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cell = "";
  let quoted = false;
  for (let i = 0; i < text.length; i += 1) {
    const char = text[i];
    const next = text[i + 1];
    if (quoted && char === '"' && next === '"') {
      cell += '"';
      i += 1;
    } else if (char === '"') {
      quoted = !quoted;
    } else if (!quoted && char === ",") {
      row.push(cell);
      cell = "";
    } else if (!quoted && (char === "\n" || char === "\r")) {
      if (char === "\r" && next === "\n") i += 1;
      row.push(cell);
      rows.push(row);
      row = [];
      cell = "";
    } else {
      cell += char;
    }
  }
  if (cell || row.length) {
    row.push(cell);
    rows.push(row);
  }
  return rows;
}

function rowsFromLocalSource(sourcePath) {
  const resolved = path.resolve(root, sourcePath);
  const text = fs.readFileSync(resolved, "utf8");
  if (resolved.endsWith(".json")) {
    const json = JSON.parse(text);
    if (Array.isArray(json.doctors)) return doctorsToRows(json.doctors);
    if (Array.isArray(json)) return json;
    throw new Error("Le fichier JSON local doit contenir un tableau ou une clé doctors.");
  }
  return parseCsv(text);
}

function doctorsToRows(doctors) {
  const header = [
    "ID", "Nom", "Prénom", "Titre", "Spécialité", "Sous-spécialité", "Ville", "Quartier",
    "Adresse", "Téléphone", "WhatsApp", "Google Maps", "Note Google", "Nombre d’avis",
    "Horaires", "Langues", "Photo", "Statut", "Vérifié", "Mis en avant", "Dernière vérification"
  ];
  const rows = doctors.map((doctor) => [
    doctor.id,
    clean(doctor.name).replace(/^Dr\\s+/i, ""),
    "",
    doctor.title || "Dr",
    doctor.specialty,
    doctor.sub_specialty,
    doctor.city,
    doctor.district,
    doctor.address,
    doctor.phone || (doctor.phones || []).join(" / "),
    doctor.whatsapp,
    doctor.google_maps,
    doctor.google_rating,
    doctor.google_reviews,
    doctor.hours,
    (doctor.languages || []).join(", "),
    doctor.image,
    doctor.status,
    doctor.verified ? "oui" : "non",
    doctor.featured ? "oui" : "non",
    doctor.updated_at
  ]);
  return [header, ...rows];
}

function readPreviousCount() {
  if (!fs.existsSync(outputPath)) return 0;
  const previous = JSON.parse(fs.readFileSync(outputPath, "utf8"));
  if (!Array.isArray(previous.doctors)) return 0;
  return previous.doctors.length;
}

function assertSafeReplacement(nextDoctors) {
  const previousCount = readPreviousCount();
  if (!previousCount || allowDoctorDrop) return;
  if (nextDoctors.length <= previousCount * 0.5) {
    throw new Error(
      `Synchronisation refusée: ${nextDoctors.length} médecins générés contre ${previousCount} existants. ` +
      "Définissez ALLOW_DOCTOR_DROP=true uniquement si cette suppression est volontaire."
    );
  }
}

async function main() {
  const rows = localSource ? rowsFromLocalSource(localSource) : await readGoogleSheet();
  const doctors = normalizeRows(rows);
  assertSafeReplacement(doctors);

  const payload = {
    updated_at: new Date().toISOString().slice(0, 10),
    source: localSource ? `local:${localSource}` : `google-sheets:${sheetRange}`,
    doctors
  };

  if (!Array.isArray(payload.doctors)) throw new Error("JSON invalide: doctors doit être un tableau.");
  const json = `${JSON.stringify(payload, null, 2)}\n`;
  JSON.parse(json);

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  const tmpPath = `${outputPath}.tmp`;
  fs.writeFileSync(tmpPath, json, "utf8");
  fs.renameSync(tmpPath, outputPath);
  console.log(`Doctors synced: ${doctors.length} fiches écrites dans ${outputPath}`);
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});

File: scripts/sync-doctors.py
#!/usr/bin/env python3
"""Import the editable specialist workbook into the site's canonical JSON data."""
import argparse, datetime, json, re, unicodedata, zipfile
from pathlib import Path
from xml.etree import ElementTree as ET

ROOT = Path(__file__).resolve().parents[1]
NS = {'m': 'http://schemas.openxmlformats.org/spreadsheetml/2006/main'}

def clean(value): return re.sub(r'\s+', ' ', str(value or '')).strip()
def norm(value):
    return ''.join(c for c in unicodedata.normalize('NFD', clean(value).lower()) if unicodedata.category(c) != 'Mn')
def slug(value):
    value = norm(value).replace('œ','oe')
    return re.sub(r'(^-|-$)', '', re.sub(r'[^a-z0-9]+', '-', value)) or 'doctor'
def phones(value): return [clean(x) for x in re.split(r'\s*[|;/]\s*', clean(value)) if clean(x)]
def specialty_group(value):
    s = norm(value)
    if 'neuro' in s: return 'Neurologie et neurochirurgie'
    if 'cardio' in s: return 'Cardiologie'
    if 'traumato' in s or 'ortho' in s: return 'Traumatologie et orthopédie'
    if 'gastro' in s: return 'Gastro-entérologie'
    if 'ophtalmo' in s: return 'Ophtalmologie'
    if 'rhumato' in s: return 'Rhumatologie'
    if 'urolog' in s: return 'Urologie'
    if 'intern' in s: return 'Médecine interne'
    return clean(value)

def xlsx_rows(path):
    with zipfile.ZipFile(path) as book:
        root = ET.fromstring(book.read('xl/sharedStrings.xml'))
        shared = [''.join(t.text or '' for t in si.iter('{%s}t' % NS['m'])) for si in root.findall('m:si', NS)]
        sheet = ET.fromstring(book.read('xl/worksheets/sheet1.xml'))
        rows = []
        for row in sheet.findall('.//m:sheetData/m:row', NS):
            values = {}
            for cell in row.findall('m:c', NS):
                ref = re.match(r'([A-Z]+)', cell.get('r', 'A1')).group(1)
                col = 0
                for char in ref: col = col * 26 + ord(char) - 64
                value = cell.find('m:v', NS)
                text = '' if value is None else (shared[int(value.text)] if cell.get('t') == 's' else value.text)
                values[col - 1] = text
            rows.append([values.get(i, '') for i in range(max(values, default=-1) + 1)])
        return rows

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('source', nargs='?', default='médecin spécialiste kenitra.xlsx')
    ap.add_argument('--allow-mass-replacement', action='store_true')
    args = ap.parse_args()
    rows = xlsx_rows(ROOT / args.source)
    header = [norm(x) for x in rows[0]]
    required = {'dr. prenom nom': 0, 'specialite': 1, 'ville': 2, 'quartier': 3, 'numero de telephone': 4, 'google maps': 5, 'instagram': 6, 'facebook': 7}
    for key, index in required.items():
        if index >= len(header) or header[index] != key: raise SystemExit(f'En-tête inattendu: colonne {index + 1}')
    doctors, current_specialty, used = [], '', set()
    for line, row in enumerate(rows[1:], 2):
        name = clean(row[0] if len(row) > 0 else '')
        if not name: current_specialty = clean(row[1] if len(row) > 1 else '') or current_specialty; continue
        specialty = clean(row[1] if len(row) > 1 else '') or current_specialty
        if not specialty: raise SystemExit(f'Ligne {line}: spécialité manquante')
        base = slug(name)
        ident = base
        if ident in used: ident = f'{base}-{slug(specialty)}'
        n = 2
        while ident in used: ident = f'{base}-{n}'; n += 1
        used.add(ident)
        raw_phone = clean(row[4] if len(row) > 4 else '')
        p = phones(raw_phone)
        doctors.append({'id': ident, 'name': name, 'title': 'Dr', 'specialty': specialty,
          'specialty_group': specialty_group(specialty), 'sub_specialty': '', 'city': clean(row[2] if len(row)>2 else '') or 'Kénitra',
          'district': clean(row[3] if len(row)>3 else ''), 'address': '', 'phone': raw_phone, 'phones': p, 'whatsapp': '',
          'google_maps': clean(row[5] if len(row)>5 else ''), 'instagram': clean(row[6] if len(row)>6 else ''),
          'facebook': clean(row[7] if len(row)>7 else ''), 'status': 'active', 'verified': False, 'featured': False})
    old_path = ROOT / 'data/doctors.json'
    old = json.loads(old_path.read_text()) if old_path.exists() else {'doctors': []}
    old_names = {norm(d.get('name')) for d in old.get('doctors', [])}
    overlap = sum(norm(d['name']) in old_names for d in doctors)
    if old_names and overlap < len(old_names) * .8 and not args.allow_mass_replacement:
        raise SystemExit(f'Remplacement refusé: {overlap}/{len(old_names)} anciens médecins retrouvés; utilisez --allow-mass-replacement après vérification.')
    name_map = {}
    phone_map = {}
    specialty_map = {}
    for d in doctors:
        keys = [name_map.setdefault(norm(d['name']), []), specialty_map.setdefault((norm(d['name']), norm(d['specialty'])), [])]
        keys[0].append(d['id']); keys[1].append(d['id'])
        for phone in d['phones']: phone_map.setdefault(re.sub(r'\D','',phone), []).append(d['id'])
    duplicates = [{'type': kind, 'key': key, 'doctor_ids': ids} for kind, mapping in [('name', name_map), ('phone', phone_map), ('name_specialty', specialty_map)] for key, ids in mapping.items() if key and len(ids) > 1]
    today = datetime.date.today().isoformat()
    payload = {'updated_at': today, 'source': f'local:{args.source}', 'doctors': doctors}
    (ROOT / 'data/doctors.json').write_text(json.dumps(payload, ensure_ascii=False, indent=2) + '\n')
    (ROOT / 'data/doctors-duplicates.json').write_text(json.dumps({'generated_at': today, 'duplicates': duplicates}, ensure_ascii=False, indent=2) + '\n')
    missing_phone = sum(not d['phones'] for d in doctors); missing_district = sum(not d['district'] for d in doctors)
    report = {'generated_at': today, 'total_doctors': len(doctors), 'specialties': len({d['specialty_group'] for d in doctors}), 'missing_phone': missing_phone, 'missing_district': missing_district, 'duplicates': len(duplicates), 'specialty_groups': sorted({d['specialty_group'] for d in doctors})}
    (ROOT / 'data/doctors-report.json').write_text(json.dumps(report, ensure_ascii=False, indent=2) + '\n')
    print(json.dumps({**report, 'overlap_with_previous': overlap}, ensure_ascii=False))
if __name__ == '__main__': main()
